File: UIImageViewModeScaleAspect/UIImageViewModeScaleAspect.m
//
//  UIImageViewModeScaleAspect.m
//
//  http://www.viviencormier.fr/
//
//  Created by Vivien Cormier on 02/05/13.
//  Copyright (c) 2013 Vivien Cormier. All rights reserved.
//
//  Extended by Alex Rupérez on 09/09/13.
//
//  http://alexruperez.com/

#import "UIImageViewModeScaleAspect.h"

@interface UIImageViewModeScaleAspect ()

@property (nonatomic, readwrite) CGRect newFrameWrapper;
@property (nonatomic, readwrite) CGRect newFrameImg;
@property (nonatomic, readwrite) CGRect lastFrameImg;
@property (nonatomic, readwrite) UIViewContentMode lastContentModeImg;
@property (nonatomic, readwrite) int lastIndexImg;
@property (nonatomic) UIImageView *img;

@end

@implementation UIImageViewModeScaleAspect

bool enlarged;
#pragma mark - Init

- (id)init {
    self = [super init];
    if (self) {
        
        self.img = [[UIImageView alloc]init];
        [self addSubview:_img];
        
        self.clipsToBounds = YES;
    }
    return self;
}
- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        
        self.img = [[UIImageView alloc]initWithFrame:[self bounds]];
        [self addSubview:_img];
        
        self.clipsToBounds = YES;
    }
    return self;
}

#pragma mark - Automatic Animate

- (void)animateToScaleAspectFitToFrame:(CGRect)frame WithDuration:(float)duration afterDelay:(float)delay{
    
    if (![self uiimageIsEmpty]) {
        _lastFrameImg = [self frame];
        _lastContentModeImg = _img.contentMode;
        _lastIndexImg = [self.superview.subviews indexOfObject:self];
        [self.superview bringSubviewToFront:self];
        enlarged = YES;
        [self initToScaleAspectFitToFrame:frame];
        
        [UIView animateWithDuration:duration delay:delay options:UIViewAnimationOptionAllowUserInteraction
                         animations:^{
                             [self animaticToScaleAspectFit];
                         } completion:^(BOOL finished) {
                             [self animateFinishToScaleAspectFit];
                         }];
    }else{
        NSLog(@"ERROR, UIImageView %@ don't have UIImage",self);
    }
    
    
}

- (void)animateToScaleAspectFillToFrame:(CGRect)frame WithDuration:(float)duration afterDelay:(float)delay{
    
    if (![self uiimageIsEmpty]) {
        enlarged = NO;
        [self initToScaleAspectFillToFrame:frame];
        
        [UIView animateWithDuration:duration delay:delay options:UIViewAnimationOptionAllowUserInteraction
                         animations:^{
                             [self animaticToScaleAspectFill];
                         } completion:^(BOOL finished) {
                             [self animateFinishToScaleAspectFill];
                         }];
    }else{
        NSLog(@"ERROR, UIImageView %@ don't have UIImage",self);
    }
    
}

- (void)animateToScaleAspectFillToTheLastFrameWithDuration:(float)duration afterDelay:(float)delay{
    [self animateToScaleAspectFillToFrame:_lastFrameImg WithDuration:duration afterDelay:delay];
}

- (void)animate{
    if (enlarged) {
        [self animateToScaleAspectFillToTheLastFrameWithDuration:0.4f afterDelay:0.0f];
    }else{
        UIWindow *window = [[[UIApplication sharedApplication] delegate] window];
        CGRect windowFrame = CGRectMake(window.bounds.origin.x, window.bounds.origin.y - 10.0f, window.bounds.size.width, window.bounds.size.height);
        if (UIInterfaceOrientationIsLandscape([[UIDevice currentDevice] orientation])){
            windowFrame = CGRectMake(window.bounds.origin.x, window.bounds.origin.y - 10.0f, window.bounds.size.height, window.bounds.size.width);
        }
        [self animateToScaleAspectFitToFrame:windowFrame WithDuration:0.4f afterDelay:0.0f];
    }
}

#pragma mark - Manual Animate

#pragma mark - - Init Function

- (void)initToScaleAspectFitToFrame:(CGRect)newFrame{
    
    if (![self uiimageIsEmpty]) {
        
        float ratioImg = (_img.image.size.width) / (_img.image.size.height);
        
        if ([self choiseFunctionWithRationImg:ratioImg ForFrame:self.frame]) {
            self.img.frame = CGRectMake( - (self.frame.size.height * ratioImg - self.frame.size.width) / 2.0f, self.frame.origin.y, self.frame.size.height * ratioImg, self.frame.size.height);
        }else{
            self.img.frame = CGRectMake(0, - (self.frame.size.width / ratioImg - self.frame.size.height) / 2.0f, self.frame.size.width, self.frame.size.width / ratioImg);
        }
    }else{
        NSLog(@"ERROR, UIImageView %@ don't have UIImage",self);
    }
    
    self.img.contentMode = UIViewContentModeScaleAspectFit;
    
    self.newFrameImg = CGRectMake(0, 0, newFrame.size.width, newFrame.size.height);
    self.newFrameWrapper = newFrame;
    
}

- (void)initToScaleAspectFillToFrame:(CGRect)newFrame{
    
    if (![self uiimageIsEmpty]) {
        
        float ratioImg = (_img.image.size.width) / (_img.image.size.height);
        
        if ([self choiseFunctionWithRationImg:ratioImg ForFrame:newFrame]) {
            self.newFrameImg = CGRectMake( - (newFrame.size.height * ratioImg - newFrame.size.width) / 2.0f, 0, newFrame.size.height * ratioImg, newFrame.size.height);
        }else{
            self.newFrameImg = CGRectMake(0, - (newFrame.size.width / ratioImg - newFrame.size.height) / 2.0f, newFrame.size.width, newFrame.size.width / ratioImg);
        }
    }else{
        NSLog(@"ERROR, UIImageView %@ don't have UIImage",self);
    }
    
    self.newFrameWrapper = newFrame;
    
}

#pragma mark - - Animatic Function

- (void)animaticToScaleAspectFit{
    
    self.img.frame = _newFrameImg;
    [self setFrameWrapper:_newFrameWrapper];
}

- (void)animaticToScaleAspectFill{
    
    self.img.frame = _newFrameImg;
    [self setFrameWrapper:_newFrameWrapper];
}

#pragma mark - - Last Function

- (void)animateFinishToScaleAspectFit{
    
}

- (void)animateFinishToScaleAspectFill{
    self.img.contentMode = _lastContentModeImg;
    self.img.frame  = CGRectMake(0, 0, self.frame.size.width, self.frame.size.height);
    UIView *superView = self.superview;
    [self removeFromSuperview];
    [superView insertSubview:self atIndex:_lastIndexImg];
}

#pragma mark - Rewrite Setter / Getter

- (void)setImage:(UIImage *)image{
    
    self.img.image = image;
    
}

- (UIImage *)image{
    
    return _img.image;
    
}

- (void)setContentMode:(UIViewContentMode)contentMode{
    
    self.img.contentMode = contentMode;
    
}

- (UIViewContentMode)contentMode{
    
    return _img.contentMode;
    
}

- (void)setFrame:(CGRect)frame{
    
    [super setFrame:frame];
    self.img.frame = frame;
}

- (void)setFrameWrapper:(CGRect)frame{
    
    [super setFrame:frame];
    
}

#pragma marf - Private Funcion

- (BOOL)uiimageIsEmpty{
    
    CGImageRef cgref = [_img.image CGImage];
    CIImage *cim = [_img.image  CIImage];
    
    if (cim == nil && cgref == NULL)
    {
        return true;
    }else{
        return false;
    }
    
}

- (BOOL)choiseFunctionWithRationImg:(float)ratioImg ForFrame:(CGRect)newFrame{
    
    BOOL resultat = false;
    
    float ratioSelf = (newFrame.size.width) / (newFrame.size.height);
    
    if (ratioImg < 1) {
        if (ratioImg > ratioSelf ) resultat = true;
    }else{
        if (ratioImg > ratioSelf ) resultat = true;
    }
    
    return resultat;
    
}

@end

File: UIImageViewModeScaleAspect/UIImageViewModeScaleAspect.swift
//
//  UIImageViewModeScaleAspect.swift
//  UIImageViewModeScaleAspect
//
//  Created by Vivien Cormier on 06/07/16.
//  Copyright © 2016 Vivien Cormier. All rights reserved.
//

import UIKit

@IBDesignable
open class UIImageViewModeScaleAspect: UIView {
    
    public enum ScaleAspect {
        case fit
        case fill
    }
    
    @IBInspectable open var image: UIImage? {
        didSet {
            transitionImage.image = image
        }
    }
    
    internal var transitionImage: UIImageView
    fileprivate var newTransitionImageFrame: CGRect?
    fileprivate var newSelfFrame: CGRect?
    
    required public init?(coder aDecoder: NSCoder) {
        
        transitionImage = UIImageView()
        transitionImage.contentMode = .center
        
        super.init(coder: aDecoder)

        addSubview(transitionImage)
        transitionImage.frame = CGRect(x: 0, y: 0, width: frame.size.width, height: frame.size.height)
        transitionImage.autoresizingMask = [ .flexibleWidth, .flexibleHeight]
        clipsToBounds = true
        
    }
    
    override public init(frame: CGRect) {
        
        transitionImage = UIImageView(frame: CGRect(x: 0, y: 0, width: frame.size.width, height: frame.size.height))
        transitionImage.contentMode = .scaleAspectFit;
        
        super.init(frame: frame)
        
        addSubview(transitionImage)
        clipsToBounds = true
        
    }
    
    //MARK: Automatic animations

    /**
     Animate the UIImageView between to UIViewContentModeScaleAspect[Fill, Fit]

     - Parameters:
        - scaleAspect: Content mode that you want to change.
        - frame: Optional parameters. New frame you want to set to your image during the animation. If nil, only the scale aspect is changed.
        - duration: The total duration of the animations, measured in seconds. If you specify a negative value or 0, the changes are made without animating them.
        - delay: Optional parameters. The amount of time (measured in seconds) to wait before beginning the animations. Specify a value of 0 to begin the animations immediately.
        - completion: Optional parameters. A block object to be executed when the animation sequence ends. This block has no return value and takes a single Boolean argument that indicates whether or not the animations actually finished before the completion handler was called. If the duration of the animation is 0, this block is performed at the beginning of the next run loop cycle. This parameter may be NULL.

     - Returns: Animated image.
     */
    open func animate(_ scaleAspect: ScaleAspect, frame: CGRect? = nil, duration: Double, delay: Double? = nil, completion: ((Bool) -> Void)? = nil) -> Void {

        var newFrame = self.frame
        if frame != nil {
            newFrame = frame!
        }

        initialeState(scaleAspect, newFrame: newFrame)

        var delayAnimation = 0.0
        if delay != nil {
            delayAnimation = delay!
        }
        
        UIView.animate(withDuration: duration, delay: delayAnimation, options: .allowAnimatedContent, animations: {
            self.transitionState(scaleAspect)
        }, completion: { (finished) in
            self.endState(scaleAspect)
            completion?(finished)
        })
        
    }
    
    //MARK: Manual animations

    /**
     If you want to animate yourself the image, you need to call this function before the animation block.

     - Parameters:
        - scaleAspect: Content mode that you want to change.

     - Returns: New frame for the image
     */
    open func initialeState(_ newScaleAspect: ScaleAspect, newFrame: CGRect) -> Void {
        
        precondition(transitionImage.image != nil)
        
        if newScaleAspect == ScaleAspect.fill && contentMode == .scaleAspectFill ||
            newScaleAspect == ScaleAspect.fit && contentMode == .scaleAspectFit {
            print("UIImageViewModeScaleAspect - Warning : You are trying to animate your image to \(contentMode) but it's already set.")
        }
        
        let ratio = transitionImage.image!.size.width / transitionImage.image!.size.height
        
        if newScaleAspect == ScaleAspect.fill {
            newTransitionImageFrame = initialeTransitionImageFrame(newScaleAspect, ratio: ratio, newFrame: newFrame)
        } else {
            transitionImage.frame = initialeTransitionImageFrame(newScaleAspect, ratio: ratio, newFrame: frame)
            transitionImage.contentMode = UIViewContentMode.scaleAspectFit;
            newTransitionImageFrame = CGRect(x: 0, y: 0, width: newFrame.size.width, height: newFrame.size.height);
        }

        newSelfFrame = newFrame
        
    }

    /**
     If you want to animate yourself the image, you need to call this function inside the animation block

     - Parameters:
        - scaleAspect: Content mode that you want to change.

     - Returns: New frame for the image
     */
    open func transitionState(_ scaleAspect: ScaleAspect) -> Void {
        transitionImage.frame = newTransitionImageFrame!
        super.frame = newSelfFrame!
    }

    /**
     If you want to animate yourself the image, you need to call this function in the completion block of the animation.

     - Parameters:
        - scaleAspect: Content mode that you want to change.

     - Returns: New frame for the image
     */
    open func endState(_ scaleAspect: ScaleAspect) -> Void {
        if scaleAspect == ScaleAspect.fill {
            transitionImage.contentMode = .scaleAspectFill;
            transitionImage.frame = CGRect(x: 0, y: 0, width: self.frame.size.width, height: self.frame.size.height);
        }
    }
    
    //MARK: Override
    
    override open var frame: CGRect {
        didSet {
            transitionImage.frame = CGRect(x: 0, y: 0, width: frame.size.width, height: frame.size.height)
        }
    }
    
    override open var contentMode: UIViewContentMode {
        get {
            return transitionImage.contentMode
        }
        set(newContentMode) {
            transitionImage.contentMode = newContentMode
        }
    }
    
    //MARK: Private
    
    fileprivate static func contentMode(_ scaleAspect: ScaleAspect) -> UIViewContentMode {
        switch scaleAspect {
        case .fit:
            return UIViewContentMode.scaleAspectFit
        case .fill:
            return UIViewContentMode.scaleAspectFill
        }
    }
    
    fileprivate func initialeTransitionImageFrame(_ scaleAspect: ScaleAspect, ratio: CGFloat, newFrame: CGRect) -> CGRect {
        
        var selectFrameFormula = false
        
        let ratioSelf = newFrame.size.width / newFrame.size.height
        
        if (ratio > ratioSelf ) {
            selectFrameFormula = true
        }
        
        if scaleAspect == ScaleAspect.fill {

            if (selectFrameFormula) {
                return CGRect( x: -(newFrame.size.height * ratio - newFrame.size.width) / 2.0, y: 0, width: newFrame.size.height * ratio, height: newFrame.size.height)
            }else{
                return CGRect(x: 0, y: -(newFrame.size.width / ratio - newFrame.size.height) / 2.0, width: newFrame.size.width, height: newFrame.size.width / ratio)
            }
            
        } else {

            if (selectFrameFormula) {
                return CGRect( x: -(frame.size.height * ratio - frame.size.width) / 2.0, y: 0, width: frame.size.height * ratio, height: frame.size.height)
            }else{
                return CGRect(x: 0, y: -(frame.size.width / ratio - frame.size.height) / 2.0, width: frame.size.width, height: frame.size.width / ratio)
            }
            
        }
        
    }
    
}
